java/ql/src/Telemetry/AutomodelFrameworkModeCharacteristics.qll

```diff
@@ -0,0 +1,333 @@
+private import semmle.code.java.dataflow.TaintTracking
+private import semmle.code.java.security.PathCreation
+private import semmle.code.java.dataflow.ExternalFlow as ExternalFlow
+private import semmle.code.java.dataflow.internal.FlowSummaryImpl as FlowSummaryImpl
+private import semmle.code.java.security.ExternalAPIs as ExternalAPIs
+private import semmle.code.java.Expr as Expr
+private import semmle.code.java.security.QueryInjection
+private import semmle.code.java.security.RequestForgery
+private import semmle.code.java.dataflow.internal.ModelExclusions as ModelExclusions
+import AutomodelSharedCharacteristics as SharedCharacteristics
+import AutomodelEndpointTypes as AutomodelEndpointTypes
+
+/**
+ * A meta data extractor. Any Java extraction mode needs to implement exactly
+ * one instance of this class.
+ */
+abstract class MetadataExtractor extends string {
+  bindingset[this]
+  MetadataExtractor() { any() }
+
+  abstract predicate hasMetadata(
+    DataFlow::ParameterNode e, string package, string type, boolean subtypes, string name,
+    string signature, int input
+  );
+}
+
+module FrameworkCandidatesImpl implements SharedCharacteristics::CandidateSig {
+  class Endpoint = DataFlow::ParameterNode;
+
+  class EndpointType = AutomodelEndpointTypes::EndpointType;
+
+  class NegativeEndpointType = AutomodelEndpointTypes::NegativeSinkType;
+
+  class RelatedLocation = Location::Top;
+
+  // Sanitizers are currently not modeled in MaD. TODO: check if this has large negative impact.
+  predicate isSanitizer(Endpoint e, EndpointType t) { none() }
+
+  RelatedLocation asLocation(Endpoint e) { result = e.asParameter() }
+
+  predicate isKnownLabel(string label, string humanReadableLabel, EndpointType type) {
+    label = "read-file" and
+    humanReadableLabel = "read file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "create-file" and
+    humanReadableLabel = "create file" and
+    type instanceof AutomodelEndpointTypes::TaintedPathSinkType
+    or
+    label = "sql" and
+    humanReadableLabel = "mad modeled sql" and
+    type instanceof AutomodelEndpointTypes::SqlSinkType
+    or
+    label = "open-url" and
+    humanReadableLabel = "open url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "jdbc-url" and
+    humanReadableLabel = "jdbc url" and
+    type instanceof AutomodelEndpointTypes::RequestForgerySinkType
+    or
+    label = "command-injection" and
+    humanReadableLabel = "command injection" and
+    type instanceof AutomodelEndpointTypes::CommandInjectionSinkType
+  }
+
+  predicate isSink(Endpoint e, string label) {
+    exists(
+      string package, string type, boolean subtypes, string name, string signature, string ext,
+      string input
+    |
+      sinkSpec(e, package, type, subtypes, name, signature, ext, input) and
+      ExternalFlow::sinkModel(package, type, subtypes, name, [signature, ""], ext, input, label, _)
```

Comment:
See https://github.com/github/CodeML/issues/113: what would be the best way to have a `regression` version of a mode that would ignore endpoints where the provenance is `ai-generated` as sinks?

Reply:
I _think_ it can be done elegantly by adding one extra characteristic and including that in the test. If not, it can be done less elegantly by modifying the framework mode's `isSink` implementation a little bit to behave differently in this specific test.